_posts/2018-02-06-unreliable-network.md

```diff
@@ -0,0 +1,116 @@
+
+Another possible improvement would be to use optimistic update: we assume that most of our network requests are going to be successful. After sending the network request we update the app to the state in which it has supposedly completed successfully, without really waiting the actual request to complete. Only in the case of failure, we reupdate view back to match the server's response. This will make the application feel more responsible to the end user, especially in cases when data is edited visually. A possible downside is that the user is not sure whether the data was actually saved or not, so perhaps the user should be informed when the data save operation has been verified to have completed.
+
+**Are you making multiple network requests to get data for a specific view?**
+
+Consider combining these requests, i.e. making a single API method to return all the data for a specific view. This way you get all the data in one query. The downside is of course that you need to wait for that one query to be complete for the view to be ready, and if that one query fails, you do not get any data at all. [GraphQL](https://en.wikipedia.org/wiki/GraphQL) is a good example of a query language which allows clients to define the exact query parameters and the structure of the data returned from the servers.
+
+**Is some specific request slowing down the initial render of your application?**
+
+Your users want your application to load as quickly as possible, and thus you should only make the network requests that are absolutely mandatory to get your application up and running. If the network infrastructure is unstable, getting the application to load itself can take a long time or even fail completely.
+
+# Retry with Increasing Timeout
+
+Now that we have optimised the network requests, it is time to make sure those requests pass to the server and back as often as possible. Application level network requests always operate on top of lower-level transmission protocols, such as [TCP](https://en.wikipedia.org/wiki/Transmission_Control_Protocol). These protocols try to provide an abstraction of a reliable connection. This is achieved by the sender detecting lost data and retransmitting it to the receiver automatically. So, if lower-level protocols already try to abstract reliable connection in every situation, why would we want to write our own retry mechanism on application level?
+
+The answer is that application level network requests, such as HTTP requests, can nominally finish, but still transiently get an error code from the other end. This can happen if there are some transient errors in proxies or load balancers. If we cannot fix these problems by making modifications to the network infrastructure, we can abstract more reliable connection on application level.
```

Comment:
~nominally~ maybe "appear to"? "nominally" means "in name only"

(~transient~ means "short-lived", but I guess that's what you actually mean here?)

"we can abstract *a* mor reliable…"